docs/DesignDocs/Enabledebugmalloc.md

```diff
@@ -0,0 +1,154 @@
+Proposal to enable debug malloc for end users
+===
+
+# Motivation
+
+Debug malloc is an internal tool which can help identify memory leaks in enclaves.
+It is turned on by default in debug builds of the SDK, and helps detect leaks in the SDK.
+
+the main logic of Debug malloc is as following:
+- Whenever memory is allocated, it captures the callstack and associates it with the allocated memory.
+  Additionally it also sets the contents of the allocated memory to zero.
+
+- Whenever memory is freed, it clears the contents to a fixed pattern to help detect access of freed memory.
+  It also frees the associated callstack information.
+
+- When enclave terminates, it prints information about leaks - those objects that have been allocated,
+  but not feed.
+
+Debug malloc is performance intensive and is therefore not compiled into the SDK release.
+
+There have been requests from users that debug malloc be made available so that they can
+identify leaks within the enclave. Commonly used memory detection tools like Valgrind do not
+work within the enclaves since they cannot access enclave memory.
+
+# Design
+
+## User Experience
+- Debug malloc will be shipped to the users as a separate library oelibdebugmalloc.a
+- To use debug malloc, user will have to link oelibdebugmalloc.a into their enclaves.
+  This can be done by
+  ```
+      cmake:
+      target_link_libraries(user-enclave oelibc oedebugmalloc.a oeenclave)
+
+      make:
+      LDFLAGS=$(shell pkg-config oeenclave-$(COMPILER) --libs)
+      LDFLAGS_CUSTOM=$(subst -loelibc, -loedebugmalloc -loelibc, $(LDFLAGS))
+      $(CC) -o enclave enc.o $(LDFLAGS_CUSTOM)
+
+  ```
+
+## Fine-grained runtime control
+Currently, debug malloc prints leaks only during enclave termination.
+However, users would like to isolate and track specific memory allocations.
+To achieve this, the following APIs are provided in a new published header `debugmalloc.h`.
+
+- `void oe_debug_malloc_start_tracking(void)`
+  Start tracking memory allocations.
+
+- `void oe_debug_malloc_stop_tracking(void)`
+  Stop tracking memory allocations.
+
+- `void oe_debug_malloc_print_objects(void)`
+  Print tracked objects that alive, ie not yet freed.
+
+- `void oe_debug_malloc_init_callback(void)`
+  This function will be called by debug malloc during startup.
+  Users can implement this function to start a tracking session at the earliest possible
+  time.
+
+Please note that the timing of calling of these functions is important. For
+example, the starting and stoping tracking must cover the coupled malloc/free
+operations, which means start tracking before malloc and also stop tracking
+after free, or debug malloc will definitely trigger false alarms. And by the same
+logic, oe_debug_malloc_print_objects() can only be called after tracking is
+stopped.
```

Comment:
In the meeting we suggested removing this restriction.

Reply:
It is removed, thanks!